src/op_analytics/cli/subcommands/pulls/agora/delegate_events.py

```diff
@@ -0,0 +1,143 @@
+from dataclasses import dataclass
+from op_analytics.cli.subcommands.pulls.agora.data_access import (
+    Agora,
+    write,
+    _camelcase_to_snakecase,
+    parse_isoformat_with_z,
+)
+from op_analytics.coreutils.logger import structlog
+from op_analytics.coreutils.request import get_data
+from op_analytics.coreutils.threads import run_concurrently_store_failures
+
+from typing import Any
+import requests
+import pandas as pd
+import os
+
+log = structlog.get_logger()
+
+BASE_URL = "https://vote.optimism.io/api/v1"
+DELEGATES_ENDPOINT = f"{BASE_URL}/delegates"
+API_KEY = os.environ["AGORA_API_KEY"]
+
+
+@dataclass
+class AgoraDelegateVotes:
+    votes_df: pd.DataFrame
+
+
+@dataclass
+class PaginatedResponse:
+    has_next: bool
+    next_offset: int
+    data: Any
+
+
+@dataclass
+class SimplePaginator:
+    url: str
+    limit: int = 50
+
+    def request(self, offset):
+        session = requests.Session()
+        result = get_data(
+            session,
+            url=self.url,
+            headers={"Authorization": f"Bearer {API_KEY}"},
+            params={"offset": offset, "limit": self.limit},
+        )
+
+        assert "meta" in result
+        assert "data" in result
+
+        return PaginatedResponse(
+            has_next=result["meta"]["has_next"],
+            next_offset=result["meta"]["next_offset"],
+            data=result["data"],
+        )
+
+    def fetch_all(self):
+        offset = 0
+        all_data = []
+
+        while True:
+            response = self.request(offset)
+            all_data.extend(response.data)
+
+            if not response.has_next:
+                break
+
+            offset = response.next_offset
+
+        return all_data
+
+
+def fetch_event_data(delegates: list, endpoint: str, workers: int) -> pd.DataFrame:
+    def fetch_delegate_data(address: str, endpoint: str) -> list:
+        paginator = SimplePaginator(url=f"{BASE_URL}/delegates/{address}/{endpoint}", limit=50)
+        try:
+            return paginator.fetch_all()
+        except requests.exceptions.HTTPError as e:
+            log.error(
+                f"HTTP Error for address: {address}, URL: {e.response.url}, Status Code: {e.response.status_code}"
+            )
+            return []
+
+    run_results = run_concurrently_store_failures(
+        function=fetch_delegate_data,
+        targets=delegates,
+        max_workers=workers,
+        function_args=(endpoint,),
+    )
+
+    all_delegate_data = run_results.results.values()
+    failed_addresses = run_results.failures.keys()
+
+    if len(failed_addresses) > 0:
+        log.error(f"Failed to fetch data for {len(failed_addresses)} addresses")
+
+    return all_delegate_data
+
+
+def _flatten_data(data):
```

Comment:
what do you think about being more explicit about the keys that you pick from values? Something like:
```
def _flatten_data(data):
    rows = []
    for record in data:
        for address, values in record.items():
            rows.append(
                {
                    "address": address,
                    "requiredX": values["requiredX"],
                    "optionalY": values.get("optionalY"),
                }
            )
    return rows 
```

I think that can help us be more aware of any possible changes in the api.

Reply:
Also if you are more explicit you can include all of your transformations in the flattening loop instead of using pandas on the overall dataframe.